📖 `bento-base-carousel` documentation (#35645)
ampproject/amphtml: extensions/amp-base-carousel/1.0/README.md

---
$category@: layout
formats:
  - websites
teaser:
  text: Displays multiple similar pieces of content along a horizontal axis or vertical axis.
---

<!---
Copyright 2021 The AMP HTML Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS-IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

# Bento Carousel

## Usage

> A generic carousel for displaying multiple similar pieces of content along a
> horizontal or vertical axis.

Reviewer: I think it may be clarifying to state that this component can be used in multiple environments upfront. Maybe:

```suggestion
A generic carousel for displaying multiple similar pieces of content along a horizontal or vertical axis. Use Bento Carousel as a web component `<bento-base-carousel>`, or a Preact/React functional component. `<BentoBaseCarousel>`.
```

Each of the Bento Carousel component’s immediate children is considered an
item in the carousel. Each of these nodes may also have arbitrary children.

The carousel consists of an arbitrary number of items, as well as optional
navigational arrows to go forward or backwards a single item.

The carousel advances between items if the user swipes or uses the customizable
arrow buttons.

> This document covers both the web component `<bento-base-carousel>` and the Preact and React functional components `<BentoBaseCarousel>`.

Reviewer: Maybe replace this with:
"Use the Bento Carousel as a standalone web component `<bento-base-carousel>` or as a Preact and React functional component `<BentoBaseCarousel>`." and link to both sections from here

Author: Added this sentence from [this thread](https://github.com/ampproject/amphtml/pull/35645#discussion_r690568800), but added the links recommended here. Thanks.

### Web Component

> The examples below demonstrate the `<bento-base-carousel>` component in standalone use.

Reviewer: What do we mean by "standalone use" - without the AMP runtime?

Author: Yes

```javascript
api.goToSlide(0); // Advance to first slide.
api.goToSlide(length - 1); // Advance to last slide.
```

##### Events

The `bento-base-carousel` API allows you to register and respond to the following events:

**slideChange**

This event is triggered when the index displayed by the carousel has changed.
The new index is available via `event.data.index`.

```javascript
carousel.addEventListener('slideChange', (e) => console.log(e.data.index))
```

#### Layout and style

Each Bento component has a small CSS library you must include to guarantee proper loading without [content shifts](https://web.dev/cls/). Because of order-based specificity, you must manually ensure that stylesheets are included before any custom styles.

```html
<link rel="stylesheet" type="text/css" href="https://cdn.ampproject.org/v0/amp-base-carousel-1.0.css">
```

Alternatively, you may also make the light-weight pre-upgrade styles available inline:

```html
<style data-bento-boilerplate>
  bento-base-carousel {
    display: block;
    overflow: hidden;
    position: relative;
  }
</style>

> ```

Reviewer: I think we need to mention this information in the first paragraph as well, since it must be included.  Maybe:

You must include each Bento component's required CSS library to guarantee proper loading and before adding custom styles. Or use the light-weight pre-upgrade styles available inline. See Layout and style.

Reviewer: Wait, is this untrue for P/React?

Author: For Preact, as long as they give explicit dimensions i.e. `width` and `height` on the component, they don't need this.

Author: Is the `You must include each Bento component's required CSS library to guarantee proper loading and before adding custom styles. Or use the light-weight pre-upgrade styles available inline. See Layout and style.` recommendation for the line to go under the `### Web component` header?

Reviewer: Thanks for the clarification - moving it under the `### Web component` header is perfect

**Container type**

The `bento-base-carousel` component has a defined layout size type. To ensure the component renders correctly, be sure to apply a size to the component and its immediate children (slides) via a desired CSS layout (such as one defined with `height`, `width`, `aspect-ratio`, or other such properties):

```css
bento-base-carousel {
  height: 100px;
  width: 100%;
}

bento-base-carousel > * {
  aspect-ratio: 4/1
}
```

> #### Behavior users should be aware of

Reviewer: Lets remove this header

> ##### RTL

Reviewer: ```suggestion
##### Right-to-left slide change
```

> `<bento-base-carousel>` currently requires that you let it know when it is in an

Reviewer: Avoid dating documentation with words like "currently".
 
```suggestion
`<bento-base-carousel>` requires that you define when it is in an
```

Author: Good idea, thanks.

right-to-left (rtl) context (e.g. Arabic, Hebrew pages). While the carousel will
generally work without this, there may be a few bugs. You can let the carousel
know that it should operate as `rtl` as follows:

```html
<bento-base-carousel dir="rtl" …>
  …
</bento-base-carousel>
```

If the carousel is in a RTL context, and you want the carousel to operate as
LTR, you can explicitly set the `dir="ltr"` on the carousel.

#### Slide layout

Slides are automatically sized by the carousel when **not** specifying
`mixed-lengths`.

```html
<bento-base-carousel …>
  <img style="height: 100%; width: 100%" src="…" />
</bento-base-carousel>
```

The slides have implicit height when the carousel is laid out.
This can easily be changed with CSS. When specifying the height,
the slide will be vertically centered within the carousel.

If you want to horizontally center your slide content, you will want to create a
wrapping element, and use that to center the content.

> ##### Number of Visible Slides

Reviewer: ```suggestion
##### Number of visible slides
```

##### Events

The `BentoBaseCarousel` API allows you to register and respond to the following events:

**onSlideChange**

This event is triggered when the index displayed by the carousel has changed.

```jsx
<BaseCarousel onSlideChange={(index) => console.log(index)}>
  <img src="puppies.jpg" />
  <img src="kittens.jpg" />
  <img src="hamsters.jpg" />
</BaseCarousel>
```

#### Layout and style

**Container type**

The `BentoBaseCarousel` component has a defined layout size type. To ensure the component renders correctly, be sure to apply a size to the component and its immediate children (slides) via a desired CSS layout (such as one defined with `height`, `width`, `aspect-ratio`, or other such properties). These can be applied inline:

```jsx
<BaseCarousel style={{width: '300px', height: '100px'}}>
  <img src="puppies.jpg" />
  <img src="kittens.jpg" />
  <img src="hamsters.jpg" />
</BaseCarousel>
```

Or via `className`:

```jsx
<BaseCarousel className='custom-styles'>
  <img src="puppies.jpg" />
  <img src="kittens.jpg" />
  <img src="hamsters.jpg" />
</BaseCarousel>
```

```css
.custom-styles {
  height: 100px;
  width: 100%;
}

.custom-styles > * {
  aspect-ratio: 4/1
}
```

> #### Behavior users should be aware of

Reviewer: Remove this heading

> ##### RTL

Reviewer: change to H4
